domain/resource/state/resource.go

```diff
@@ -407,198 +421,604 @@ WHERE uuid = $resourceIdentity.uuid`,
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Check unit resource is not already inserted.
+		err := tx.Query(ctx, checkUnitResourceStmt, unitResourceInput).Get(&unitResourceInput)
+		if err == nil {
+			// If the unit to resource link is already there, return.
+			return nil
+		} else if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
 			return errors.Capture(err)
 		}
 
-		// Verify if the application is already retrieved.
-		err = tx.Query(ctx, checkAlreadyRetrievedStmt, resourceUUID).Get(&resourceUUID)
-		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		// Check resource exists.
+		err = tx.Query(ctx, checkResourceExistsStmt, unitResourceInput).Get(&unitResourceInput)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("resource %s: %w", resourceUUID, resourceerrors.ResourceNotFound)
+		} else if err != nil {
 			return errors.Capture(err)
 		}
 
-		// Update retrieved by if it is not retrieved.
+		// Check unit exists.
+		err = tx.Query(ctx, checkValidUnitStmt, unitResourceInput).Get(&unitResourceInput)
 		if errors.Is(err, sqlair.ErrNoRows) {
-			err = tx.Query(ctx, getRetrievedByTypeStmt, retrievedTypeParam).Get(&retrievedTypeParam)
-			if errors.Is(err, sqlair.ErrNoRows) {
-				return resourceerrors.UnknownRetrievedByType
-			}
-			if err != nil {
-				return errors.Capture(err)
-			}
+			return errors.Errorf("unit %s: %w", unitUUID, resourceerrors.UnitNotFound)
+		} else if err != nil {
+			return errors.Capture(err)
+		}
 
-			// Insert retrieved by.
-			retrievedByParam.RetrievedByTypeID = retrievedTypeParam.ID
-			err = tx.Query(ctx, insertRetrievedByStmt, retrievedByParam).Run()
-			if err != nil {
-				return errors.Capture(err)
-			}
+		// Unset any existing resources with the same charm resource as the
+		// resource being set in the unit resource table.
+		err = st.unsetUnitResourcesWithSameCharmResource(ctx, tx, resourceUUID)
+		if err != nil {
+			return errors.Errorf(
+				"removing previously set unit resources for resource %s: %w", resourceUUID, err,
+			)
 		}
 
-		// update unit resource table.
+		// Update unit resource table.
 		err = tx.Query(ctx, insertUnitResourceStmt, unitResourceInput).Run()
 		return errors.Capture(err)
 	})
 
-	return resource.SetUnitResourceResult{
-		UUID:      coreresource.UUID(unitResourceInput.ResourceUUID),
-		Timestamp: unitResourceInput.AddedAt,
-	}, err
+	return err
 }
 
-// OpenApplicationResource returns the metadata for a resource.
-func (st *State) OpenApplicationResource(
-	ctx context.Context,
-	resourceUUID coreresource.UUID,
-) (resource.Resource, error) {
-	return resource.Resource{}, nil
-}
+// unsetUnitResourcesForCharmResource removes all unit resources that use a
+// charm resource.
+func (st *State) unsetUnitResourcesWithSameCharmResource(
+	ctx context.Context, tx *sqlair.TX, uuid coreresource.UUID) error {
```

Comment:
todo: You need to include the unit uuid in this equation. 

Think about the scenario where there are two application with the same charm. One application upgrades with a new resource and the other doesn't. This method could be removing data on both applications.

Another scenario is three units upgrading their resource, you'll remove the 2nd two units data when the first unit updates. There is no guarantee the 2nd two are successful.

Reply:
Ah, I can't believe I missed this! Have added it in.